### PR TITLE
os/bluestore/NVMEDevice: Add multiple thread support for SPDK I/O thread

### PR DESCRIPTION
**Pull request by optimistyzy**
Previously, we only have one thread to do SPDK I/O, and this patch
adds the multiple support

Signed-off-by: Ziye Yang <optimistyzy@gmail.com>

### CODE REVIEW

**Comment by optimistyzy** [comment]
@yuyuyu101 I re-opened this PR.   @liuhongtong helped me a lot to test the functionality of multiple threads.  And this patch will not affect the usage of the previous single thread.  Also, we will dig more later for the performance of multiple I/O threads.

**Comment by liupan1111** [comment]
@optimistyzy could you say something about the performance enhancement of this multi thread?

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -16,6 +16,8 @@
  */
 
 #include <unistd.h>
+#include <sys/syscall.h>
+#define gettid() syscall(SYS_gettid)
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: why gettid instead pthread_self()

**Comment by optimistyzy** [comment]
We  use pthread_self() first, after the test, we did not get the
identification we ant, and use gettid is more accurate. If use
pthread_self(),  all the id info is even number. Also compare the gettid
with the info, it is the info what we want.

2017-04-11 9:14 GMT+08:00 Haomai Wang <notifications@github.com>:

File: src/os/bluestore/NVMEDevice.cc (hunk 1 above)
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: common/io_priority.h
20:extern pid_t ceph_gettid();

**Comment by yuyuyu101** [comment]
I think the sharding way should follow bluestore hash instead of thread id.

**Comment by yuyuyu101** [comment]
@liewegas any idea on future sharding model?

**Comment by optimistyzy** [comment]
@yuyuyu101 can we firstly use thread_id.  Currently, We use fio in the client for test.  Via thread id,  we can distribute all the I/Os from one thread to the dedicated SPDK I/O thread, which seems a natural way.

**Comment by liewegas** [comment]
Seems okay for now.  Agree that eventually we should use the bluestore/osd sharding model though!

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -1058,7 +1133,9 @@ int NVMEDevice::read_random(uint64_t off, uint64_t len, char *buf, bool buffered
     t->copy_to_buf(buf, off-t->offset, len);
   };
   ++ioc.num_reading;
-  driver->queue_task(t);
+  if(queue_id == -1)
+        queue_id = ceph_gettid();
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: indentation and whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -992,7 +1067,9 @@ int NVMEDevice::read(uint64_t off, uint64_t len, bufferlist *pbl,
     t->copy_to_buf(buf, 0, t->len);
   };
   ++ioc->num_reading;
-  driver->queue_task(t);
+  if(queue_id == -1)
+        queue_id = ceph_gettid();
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: indentation and whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 4]
```diff
@@ -954,7 +1027,9 @@ int NVMEDevice::aio_write(
 
   if (buffered) {
     // Only need to push the first entry
-    driver->queue_task(t);
+    if(queue_id == -1)
+	queue_id = ceph_gettid();
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: indentation and whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 5]
```diff
@@ -926,7 +997,9 @@ void NVMEDevice::aio_submit(IOContext *ioc)
     ioc->num_pending -= pending;
     assert(ioc->num_pending.load() == 0);  // we should be only thread doing this
     // Only need to push the first entry
-    driver->queue_task(t, pending);
+    if(queue_id == -1)
+	queue_id = ceph_gettid();
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: indentation and whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 6]
```diff
@@ -907,10 +973,15 @@ int NVMEDevice::flush()
 {
   dout(10) << __func__ << " start" << dendl;
   auto start = ceph::coarse_real_clock::now();
-  driver->flush_wait();
+
+  if(queue_id == -1)
+	queue_id = ceph_gettid();
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: indentation and whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 7]
```diff
@@ -755,17 +820,18 @@ void io_complete(void *t, const struct spdk_nvme_cpl *completion)
...
+  ++queue->completed_op_seq;
   auto dur = std::chrono::duration_cast<std::chrono::nanoseconds>(
       ceph::coarse_real_clock::now() - task->start);
   if (task->command == IOCommand::WRITE_COMMAND) {
-    driver->logger->tinc(l_bluestore_nvmedevice_aio_write_lat, dur);
+    queue->logger->tinc(l_bluestore_nvmedevice_aio_write_lat, dur);
     assert(!spdk_nvme_cpl_is_error(completion));
     dout(20) << __func__ << " write/zero op successfully, left "
-             << driver->queue_op_seq - driver->completed_op_seq << dendl;
+             <<queue->queue_op_seq - queue->completed_op_seq << dendl;
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: whitespace

File: src/os/bluestore/NVMEDevice.cc [hunk 8]
```diff
@@ -500,10 +562,13 @@ void SharedDriverData::_aio_thread()
           flush_cond.Signal();
       }
 
-
       if (!inflight) {
-        for (auto &&it : registered_devices)
-          it->reap_ioc();
+        //be careful, here we need to have a methods,and let each thread reap its own, currently it is done
+        // by only one dedicatd dpdk thread
+        if(!queueid) {
```
Comment (liewegas) [inline src/os/bluestore/NVMEDevice.cc]: whitespace

**Comment by optimistyzy** [comment]
@liewegas  updated with your comments.

**Comment by optimistyzy** [comment]
@yuyuyu101 @liewegas can this patch be merged？

**Comment by yuyuyu101** [comment]
retest this please

**Comment by optimistyzy** [comment]
Already tested..

2017-04-13 8:37 GMT+08:00 Haomai Wang <notifications@github.com>:

**Comment by tchaikov** [comment]
@optimistyzy "r3test this please" is a command to instruct jenkins to re-trigger the "make check" test.

**Comment by optimistyzy** [comment]
@tchikov  thanks for explaining.  It seems that I misunderstand the meaning.

**Pull request merged**